proxy/src/control/cache.rs

```diff
@@ -93,106 +133,151 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
         self.reset_on_next_modification = false;
     }
 
-    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
-        self.retain(&HashSet::new(), on_change)
+    pub fn clear<F>(&mut self, on_change: &mut F)
+    where
+        F: FnMut((K, V), CacheChange),
+    {
+        self.update_intersection(IndexMap::new(), on_change)
     }
 
-    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
-        where F: FnMut(T, CacheChange)
+    /// Update the cache to contain the intersection of its current contents
+    /// and the key-value pairs in `to_update`. Pairs not present in
+    /// `to_update` will be removed from the cache, and any keys in the cache
+    /// with different inner from those in `to_update` will be ovewritten to
+    /// match `to_update`.
+    pub fn update_intersection<F, Q>(
+        &mut self,
+        mut to_update: IndexMap<Q, V>,
+        mut on_change: F
+    )
+    where
+        F: FnMut((K, V), CacheChange),
+        K: Borrow<Q>,
+        Q: Hash + Eq,
     {
-        self.values.retain(|value| {
-            let retain = to_retain.contains(&value);
-            if !retain {
-                on_change(*value, CacheChange::Removal)
+        self.inner.retain(|key, value| {
+            match to_update.remove(key.borrow()) {
+                // New value matches old value. Do nothing.
+                Some(ref new_value) if new_value == value => true,
+                // If the new value isn't equal to the old value, overwrite
+                // the old value.
+                Some(new_value) =>  {
+                    let old_value = mem::replace(value, new_value);
+                    on_change((*key, old_value), CacheChange::Modification);
```

Comment:
Sorry I didn't notice this before: We need to send the *new* value in CacheChange::Modification, not the old value, right?

Reply:
Yes, you're right --- thanks for catching that!

Reply:
Fixed in dfe8a1febb78ed54700aeee775a6d2b1f305831a